### PR TITLE
JSON: replace YAJL with RapidJSON

### PR DESCRIPTION
**Pull request by Montellese**
This is an approach at replacing our JSON parsing/writing done with YAJL with RapidJSON which, according to a lot of benchmarks (https://code.google.com/p/rapidjson/wiki/Performance, https://github.com/mloskot/json_benchmark, https://github.com/miloyip/nativejson-benchmark, http://chadaustin.me/2013/01/json-parser-benchmarking/) is one of the fastest JSON libraries available and a lot faster than YAJL. We are not likely to notice this a lot because most of the JSON we process is rather small but it certainly won't hurt either.

I've also written a few unit tests for the JSON parsing/writing (which are independent of the JSON library we use).

rapidjson doesn't seem to be available for Ubuntu so it would need to be added to our PPA. I also haven't uploaded the source tarball and the win32 dependency archive yet.

TODOs:
- [x] upload `rapidjson-1.0.2.tar.gz` to mirrors
- [x] upload `rapidjson-1.0.2-win32.7z` to mirrors
- [x] add `rapidjson` to PPA

### CODE REVIEW

File: xbmc/utils/test/TestJSONVariantWriter.cpp [hunk 1]
```diff
@@ -23,11 +23,142 @@
...
+  variant.push_back("foo");
+  ASSERT_TRUE(CJSONVariantWriter::Write(variant, str, false));
+  ASSERT_STREQ("[\n\ttrue,\n\t\"foo\"\n]", str.c_str());
+
+  variant.clear();
+  CVariant obj(CVariant::VariantTypeObject);
+  obj["foo"] = "bar";
+  variant.push_back(obj);
+  ASSERT_TRUE(CJSONVariantWriter::Write(variant, str, false));
+  // TODO: ASSERT_TRUE(CJSONVariantWriter::Write("[ { \"foo\": \"bar\" } ]", variant));
```
Comment (Razzeee) [inline xbmc/utils/test/TestJSONVariantWriter.cpp]: Forgot about this?

**Comment by Paxxi** [comment]
As it's a header only library can't we just stick this in contrib instead of handling it through depends?

**Comment by Montellese** [comment]
@Paxxi: Personally I'd prefer not to put any 3rd party code into our repo.

File: xbmc/utils/JSONVariantParser.cpp [hunk 2]
```diff
@@ -1,214 +1,233 @@
...
 
-  m_status = ParseVariable;
-}
+  CVariant& m_parsedObject;
+  std::vector<CVariant *> m_parse;
+  std::string m_key;
+
+  enum class PARSE_STATUS
+  {
+    Variable = 0,
```
Comment (Paxxi) [inline xbmc/utils/JSONVariantParser.cpp]: No need to set values. Not even sure it's legal?

**Comment by Paxxi** [comment]
I'm split on the topic but we have started down that path with kissfft.

Code is looking good and I'm thinking with a bit of testing and the depends stuff taken care of it's good to merge.

**Comment by Montellese** [comment]
Adressed the comments. Thanks for the feedback.

**Comment by Montellese** [comment]
I've added an extra commit that enables SSE2 optimization on WIN32. Unfortunately this has to be specified at compile time and doesn't support runtime detection. But on WIN32 SSE2 support is mandatory. I have no idea about other platforms.

**Comment by Tolriq** [comment]
JSON encoding speed was a major problem on low end devices for syncing of database for remotes.

This needed to split queries to small ones and sending multiple request to webserver.

This was not optimal as this means also multiple queries to database with limits that also adds some slowness.

I'd be glad to test and report on those large json queries (multiple MB on large libraries) on Windows and rpi if @MilhouseVH can build the images.

**Comment by Montellese** [comment]
I've uploaded the rapidjson packages to our mirrors but it's gonna take an hour or so for them to be available. Then I can at least trigger a windows test build for you @Tolriq.

**Comment by stefansaraev** [comment]
@Montellese fyi. rapidjson got a hardcoded -march=native in it's CMakeLists.txt. you need something like http://sprunge.us/HZbS

**Comment by Montellese** [comment]
@stefansaraev: Actually I wrote a patch for that in rapidjson's depends integration but looks like I forgot to push that. Will do ASAP.

**Comment by Montellese** [comment]
@Tolriq: WIN32 test build is running: http://jenkins.kodi.tv/job/WIN-32/6116/

**Comment by Tolriq** [comment]
@Montellese 
Ok will be a little late for today I think, will try to check this week end, but for sure Monday, as speed gain on this would be great.

Any specific things that can have side effect ? Like null handling or empty arrays ?

**Comment by Montellese** [comment]
Theoretically it should be identical in behaviour.

**Comment by MilhouseVH** [comment]
@stefansaraev has kindly provided me with a patch to add RapidJSON to OE (adding new packages really isn't my thing!) - if all goes well this PR will be included in tonight's build. What's the best way to test the change, just call the existing JSON API methods?

**Comment by Montellese** [comment]
@MilhouseVH: cool thanks. Yeah just use the JSON-RPC API.

**Comment by MilhouseVH** [comment]
Can this be rebased now that #8018 has merged?

**Comment by Montellese** [comment]
@MilhouseVH: done. jenkins build this please

**Comment by wsnipex** [comment]
whats the required min version? ubuntu >=15.04 has a (rather old) package: https://launchpad.net/ubuntu/+source/rapidjson

**Comment by Montellese** [comment]
@wsnipex: I've downloaded the latest release v1.0.2 but I didn't check any previous releases. I have no clue what state version 0.12 is in that is available on Ubuntu. I was only able to find 0.11 and 1.0 so not sure what they are doing there.

**Comment by Tolriq** [comment]
@Montellese After quite some test all seems to work perfectly.

For speed well all my windows machine are quite fast so even on 12MB JSON files there's no difference between this and normal Jarvis build.

Only bug I get was invalid json data when asking for getEpisodes during a scrape, but I'm pretty sure this is a more global bug and not related to this PR.

**Comment by wsnipex** [comment]
strange "lib", headers only :) @Montellese temporary location, till verified: https://launchpad.net/~wsnipex/+archive/ubuntu/kodi-git/+packages?field.name_filter=rapidjson&field.status_filter=published&field.series_filter=

**Comment by L-S-D** [comment]
since this PR is included in Milhouses OE test builds, I'm no longer able to list channels in Kodi Remote for IOS. I get the following JSON error in the app:

XBMC JSON-RPC Error: Invalid params.
(Code: -32602)
Data: (null)

METHOD
PVR.GetChannels

PARAMETERS
{channelgroupid=5;properties=(thumbnail,channel);}

**Comment by MilhouseVH** [comment]
Yes, `PVR.GetChannels` is definitely a problem with this change.

You can query TV and Radio channels with my `texturecache.py` script, eg. `texturecache.py jd pvr.tv` or `texturecache.py jd pvr.tv` when using the PVR Demo client. The script queries the relevant `PVR.GetChannelGroups` method and then iterates over each group calling `PVR.GetChannels`.

With YAJL the correct details are returned for all TV and Radio channels.

With RapidJSON the above error is returned whenever `PVR.GetChannels` is queried.

Request:
```
{"jsonrpc": "2.0", "params": {"channelgroupid": 1, "properties": ["channeltype", "channel", "thumbnail", "hidden", "locked", "lastplayed"], "limits": {"start": 0, "end": 400}}, "method": "PVR.GetChannels", "id": "libPvr.radio.channel"}
```
RapidJSON Response:
```
{"error":{"code":-32602,"message":"Invalid params."},"id":"libPvr.radio.channel","jsonrpc":"2.0"}
```

YAJL Response:
```
{"id":"libPvr.radio.channel","jsonrpc":"2.0","result":{"channels":[{"channel":"Demo Radio Channel 1","channelid":25,"channeltype":"radio","hidden":false,"label":"Demo Radio Channel 1","lastplayed":"1970-01-01","locked":false,"thumbnail":"image://special%3a%2f%2fxbmc%2faddons%2fpvr.demo%2fdata%2f01.png/"},{"channel":"Demo Radio Channel 2","channelid":26,"channeltype":"radio","hidden":false,"label":"Demo Radio Channel 2","lastplayed":"1970-01-01","locked":false,"thumbnail":"image://special%3a%2f%2fxbmc%2faddons%2fpvr.demo%2fdata%2f02.png/"},{"channel":"Demo Radio Channel 3","channelid":27,"channeltype":"radio","hidden":false,"label":"Demo Radio Channel 3","lastplayed":"1970-01-01","locked":false,"thumbnail":"image://special%3a%2f%2fxbmc%2faddons%2fpvr.demo%2fdata%2f03.png/"},{"channel":"Demo Radio Channel 4","channelid":28,"channeltype":"radio","hidden":false,"label":"Demo Radio Channel 4","lastplayed":"1970-01-01","locked":false,"thumbnail":"image://special%3a%2f%2fxbmc%2faddons%2fpvr.demo%2fdata%2f04.png/"},{"channel":"Demo Radio Channel 5","channelid":29,"channeltype":"radio","hidden":false,"label":"Demo Radio Channel 5","lastplayed":"1970-01-01","locked":false,"thumbnail":"image://special%3a%2f%2fxbmc%2faddons%2fpvr.demo%2fdata%2f05.png/"},{"channel":"Demo Radio Channel 6","channelid":30,"channeltype":"radio","hidden":false,"label":"Demo Radio Channel 6","lastplayed":"1970-01-01","locked":false,"thumbnail":"image://special%3a%2f%2fxbmc%2faddons%2fpvr.demo%2fdata%2f06.png/"},{"channel":"Demo Radio Channel 7","channelid":31,"channeltype":"radio","hidden":false,"label":"Demo Radio Channel 7","lastplayed":"1970-01-01","locked":false,"thumbnail":"image://special%3a%2f%2fxbmc%2faddons%2fpvr.demo%2fdata%2f07.png/"},{"channel":"Demo Radio Channel 8","channelid":32,"channeltype":"radio","hidden":false,"label":"Demo Radio Channel 8","lastplayed":"1970-01-01","locked":false,"thumbnail":"image://special%3a%2f%2fxbmc%2faddons%2fpvr.demo%2fdata%2f08.png/"},{"channel":"Demo Radio Channel 9","channelid":33,"channeltype":"radio","hidden":false,"label":"Demo Radio Channel 9","lastplayed":"1970-01-01","locked":false,"thumbnail":"image://special%3a%2f%2fxbmc%2faddons%2fpvr.demo%2fdata%2f09.png/"},{"channel":"Demo Radio Channel 10","channelid":34,"channeltype":"radio","hidden":false,"label":"Demo Radio Channel 10","lastplayed":"1970-01-01","locked":false,"thumbnail":"image://special%3a%2f%2fxbmc%2faddons%2fpvr.demo%2fdata%2f10.png/"}],"limits":{"end":10,"start":0,"total":10}}}
```

**Comment by Montellese** [comment]
@L-S-D @MilhouseVH Thanks for the reports. I think I figured out the issue. YAJL considers most integers as signed integers whereas RapidJSON considers them as unsigned integers (as long as they aren't negative). Will see how to fix it.

**Comment by Montellese** [comment]
@MilhouseVH rebased and added a commit that should fix the issue.
@topfs2 mind taking a look since you introduced YAJL and `CVariant`? I had to adjust `CVariant::isInteger()` (which always confused me a bit).

**Comment by MilhouseVH** [comment]
Thanks, that seems to be fixed now. I've queried pvr tv and radio channels, artists, albums, songs, movies, tvshows, sets and addons and they're all working as expected now. Hopefully there's no other gotchas elsewhere... difficult to test every method!

**Comment by MilhouseVH** [comment]
This has been OK for quite a while now - time to merge?

**Comment by MartijnKaijser** [comment]
@MilhouseVH nope as merge window ended and we are going beta. Will have to wait for v17

**Comment by Razzeee** [comment]
Bump, @Montellese rebase and merge?

**Comment by Montellese** [comment]
Rebased. I'm not sure what needs to be done for this to work with PPA though. @wsnipex has setup a temporary rapidjson project on his PPA but I don't know if it has ever been tested.
jenkins build this please

**Comment by MartijnKaijser** [comment]
@montellese perhaps now the massive input PR has been merged this could get included? Was only the PPA package needed? Ping @wsnipex

**Comment by wsnipex** [comment]
its in the ppa for months now.

**Comment by Montellese** [comment]
@wsnipex: I wasn't sure if anyone ever tested it and whether it was only a temporary approach/solution or not because you said

> temporary location, till verified

**Comment by wsnipex** [comment]
right, nothing has changed in that regard.

**Comment by DaveTBlake** [comment]
@Montellese could this be rebased for v18 and completed?

Just discovered that too rapid a series of JSON commands each needing a large amount of data processing, although data not necessarily returned in the final results, can resullt in a memory leak. So say several calls to AudioLibrary.GetSongs, the 2nd call sent before the 1st completes etc.. Only produced on Pi with MySQL so far, but wondering if YAJL contributes in any way.

**Comment by AchimTuran** [comment]
@Montellese
Could you rebase this PR, then I will try to add the required CMake code next week and we could drop the project files from this PR for Windows, Linux...

Or we could ask @fetzerch or @hudokkow for CMake help if they have the time and motivation.

**Comment by AchimTuran** [comment]
Updating the rapidjson library could be done in a new PR as the version added by @Montellese should be available on our mirrors and ppa.

**Comment by hudokkow** [comment]
Rebase the branch and I'll have a go at it.

**Comment by Montellese** [comment]
Rebased but obviously completely untested since I can't build anymore without cmake integration. @hudokkow thanks for looking into it.

**Comment by hudokkow** [comment]
Thx!

**Comment by hudokkow** [comment]
Quick and dirty. Top commit in https://github.com/hudokkow/xbmc/commits/PR8008

Not working on Zesty with http://packages.ubuntu.com/zesty/rapidjson-dev

> [ 47%] Building CXX object build/addons/CMakeFiles/addons.dir/AddonDatabase.cpp.o
/home/hudo/dev/kodiws/kodi/xbmc/addons/AddonDatabase.cpp: In function ‘std::__cxx11::string SerializeMetadata(const ADDON::IAddon&)’:
/home/hudo/dev/kodiws/kodi/xbmc/addons/AddonDatabase.cpp:79:49: error: no matching function for call to ‘CJSONVariantWriter::Write(CVariant&, bool)’
   return CJSONVariantWriter::Write(variant, true);
                                                 ^
In file included from /home/hudo/dev/kodiws/kodi/xbmc/addons/AddonDatabase.cpp:32:0:
/home/hudo/dev/kodiws/kodi/xbmc/utils/JSONVariantWriter.h:31:15: note: candidate: static bool CJSONVariantWriter::Write(const CVariant&, std::__cxx11::string&, bool)
   static bool Write(const CVariant &value, std::string& output, bool compact);
               ^~~~~
/home/hudo/dev/kodiws/kodi/xbmc/utils/JSONVariantWriter.h:31:15: note:   candidate expects 3 arguments, 2 provided
/home/hudo/dev/kodiws/kodi/xbmc/addons/AddonDatabase.cpp: In function ‘void DeserializeMetadata(const string&, ADDON::CAddonBuilder&)’:
/home/hudo/dev/kodiws/kodi/xbmc/addons/AddonDatabase.cpp:84:56: error: no matching function for call to ‘CJSONVariantParser::Parse(const string&)’
   CVariant variant = CJSONVariantParser::Parse(document);
                                                        ^
In file included from /home/hudo/dev/kodiws/kodi/xbmc/addons/AddonDatabase.cpp:31:0:
/home/hudo/dev/kodiws/kodi/xbmc/utils/JSONVariantParser.h:31:15: note: candidate: static bool CJSONVariantParser::Parse(const char*, CVariant&)
   static bool Parse(const char* json, CVariant& data);
               ^~~~~
/home/hudo/dev/kodiws/kodi/xbmc/utils/JSONVariantParser.h:31:15: note:   candidate expects 2 arguments, 1 provided
/home/hudo/dev/kodiws/kodi/xbmc/utils/JSONVariantParser.h:32:15: note: candidate: static bool CJSONVariantParser::Parse(const string&, CVariant&)
   static bool Parse(const std::string& json, CVariant& data);
               ^~~~~
/home/hudo/dev/kodiws/kodi/xbmc/utils/JSONVariantParser.h:32:15: note:   candidate expects 2 arguments, 1 provided
build/addons/CMakeFiles/addons.dir/build.make:134: recipe for target 'build/addons/CMakeFiles/addons.dir/AddonDatabase.cpp.o' failed
make[2]: *** [build/addons/CMakeFiles/addons.dir/AddonDatabase.cpp.o] Error 1
CMakeFiles/Makefile2:5074: recipe for target 'build/addons/CMakeFiles/addons.dir/all' failed
make[1]: *** [build/addons/CMakeFiles/addons.dir/all] Error 2
Makefile:160: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by Montellese** [comment]
Thanks @hudokkow. That build error looks like a new usage of the JSON parser which wasn't there when I wrote the initial PR. Will fix it.

**Comment by Montellese** [comment]
jenkins build this please

**Comment by Montellese** [comment]
@hudokkow sorry about that. I removed the @ from the commit message. Or should I remove your name completely? jenkins build this please

**Comment by hudokkow** [comment]
I'm fine either way. Your call. :wink:

**Comment by MilhouseVH** [comment]
This is now building fine, thanks.

However the Ubuntu 16.04 version of RapidJSON is horribly out of date (appears to be based on RapidJSON v0.11, which is a pre-release version) - could this dependency be installed the same way as `fmt` and `spdlog`, from mirrors.kodi.tv into `/usr/local`? Perhaps then the latest v1.1.0 version could be used (although 1.1.0 currently fails to compile: http://pastebin.com/raw/3a34Ytgi).

**Comment by Montellese** [comment]
@MilhouseVH thanks for testing. @wsnipex has already prepared a newer PPA build, see https://launchpad.net/~wsnipex/+archive/ubuntu/kodi-git/+packages?field.name_filter=rapidjson&field.status_filter=published&field.series_filter= which will be integrated into the kodi build-deps PPA. But if needed I can modify the depends `Makefile` to also allow manual building on native linux builds. Should I do that (e.g. for non-ubuntu linuxes)?

**Comment by hudokkow** [comment]
Build is fine but fails with tests.

**Comment by Montellese** [comment]
Well actually one of the new tests doesn't build but only on OSX. Not sure why it doesn't work there but everywhere else.

**Comment by Montellese** [comment]
jenkins build this please

**Comment by MilhouseVH** [comment]
I'll add this to my test builds starting tonight.

**Comment by Montellese** [comment]
Great now the test compiles on OSX64 and fails on linux (with the same error message as previously on OSX64) :-/

**Comment by Montellese** [comment]
One more try. jenkins build this please

**Comment by Montellese** [comment]
Finally :-)

Now the only question left is whether I should pimp the `rapidjson` depends `Makefile` to support native linux building. @wsnipex any opinion?

**Comment by MilhouseVH** [comment]
Any chance of a quick rebase after #11072?

**Comment by Montellese** [comment]
@MilhouseVH: done

**Comment by MrMC** [comment]
double plus 1 on this. We use json heavy in upcoming native emby client. Some timing numbers;
yajl:
CEmbyUtils::GetEmbyCVariant parsed in 6235(msec)

rapidjson:
CEmbyUtils::GetEmbyCVariant parsed in 256(msec)

So parsing time becomes mostly irrelevant with rapidjson.

Memory Leak testing is the next step.

**Comment by Montellese** [comment]
@MrMC: Thanks for the numbers. I only ran very primitive tests on usual JSON-RPC requests where there was no difference in timing but then again the timings were below 10ms so I could only make sure that it wasn't worse than before.

I'm still waiting on feedback whether I should adjust the depends Makefile to also support native linux builds.

**Comment by wsnipex** [comment]
Yes, please add the native build.

**Comment by MrMC** [comment]
might diddle the depends makefile to not install the docs/examples into depends

**Comment by Montellese** [comment]
@MrMC: that's odd I've added cmake flags to disable docs, examples and tests. So it should not build/install those unless I did something wrong.

**Comment by Montellese** [comment]
I made the changes to the depends `Makefile` but wasn't able to test them right now. jenkins build this please

**Comment by MrMC** [comment]
http://pastebin.com/Tg14aV3m using this Makefile (http://pastebin.com/yF20fVmC)

**Comment by Montellese** [comment]
@MrMC: That's odd I'd have expected the `-DRAPIDJSON_BUILD_DOC=OFF -DRAPIDJSON_BUILD_EXAMPLES=OFF` flags to avoid installing those files. But maybe it's a bug in the cmake project of RapidJSON.

**Comment by stefansaraev** [comment]
-DRAPIDJSON_BUILD_DOC=OFF -DRAPIDJSON_BUILD_EXAMPLES=OFF are about "building" docs and they do as said.
 
https://github.com/miloyip/rapidjson/blob/master/CMakeLists.txt#L164-L171

just ignore it and merge this ;)

**Comment by MrMC** [comment]
or add a rm hammer after the install

rm -rf $(PREFIX)/share/doc/RapidJSON

that will nuke them

**Comment by Montellese** [comment]
@MrMC: added.

**Comment by Montellese** [comment]
jenkins build this please one last time

**Comment by wsnipex** [comment]
please add rapidjson-dev to https://github.com/xbmc/xbmc/blob/master/.travis.yml#L71-L79

**Comment by Montellese** [comment]
Is this ok to merge or do we want to update to RapidJSON 1.1.0 (latest release) instead? Apart from fixes and a lot of new features that we probably don't need it comes with C++11 and JSON schema support. The latter might be useful to try and replace our own JSON schema parser for JSON-RPC. But obviously that would be out of scope for this PR.

**Comment by Montellese** [comment]
@wsnipex: done

**Comment by AchimTuran** [comment]
@montellese
I guess we can update when somebody replaces our own json scheme parser and then as you said create a new PR.

Thanks to all that were involved here to add rapidjson to Kodi. Very cool library that improves a lot and helps during AudioDSP V2 development.

**Comment by Paxxi** [comment]
I think we should bump the version to latest but that's best done in another pr.

**Comment by Razzeee** [comment]
+1 Paxxi

**Comment by pgnd** [comment]
since this has apparently been committed, and the dev branch closed, not entirely sure where this comment should go ...

building kodi/head

```
git log | head -n1
    commit 78cce5735bc5a67e7a448bf18c5ac6b1d0ca2326
```
on linux64, my latest make fails @
```
/usr/include/rapidjson/reader.h:1340:32: error: ‘class CJSONVariantParserHandler’ has no member named ‘RawNumber’
                  cont = handler.RawNumber(str, SizeType(length), false);
```
bit more at

https://pastebin.com/4B9nAiFN

**Comment by MilhouseVH** [comment]
@pgnd see https://github.com/xbmc/xbmc/pull/8008#issuecomment-285632657 - looks like you're building with RapidJSON v1.1.0 not v1.0.2.

**Comment by stefansaraev** [comment]
please test https://github.com/stefansaraev/xbmc/commit/b89061cda677876ae562f1e00657ccd12cac2821 and let me know if it works.

**Comment by MilhouseVH** [comment]
Thanks @stefansaraev, your commit and v1.1.0 builds and seems to runs fine with LibreELEC/Linux.

**Comment by candrews** [comment]
https://github.com/xbmc/xbmc/pull/8008#issuecomment-288858221 Gentoo user reporting in :)

So is there a plan to support rapidjson 1.1.0? It's going to be awkward as distributions upgrade rapidjson to 1.1.0 yet kodi won't build with it.

**Comment by Rechi** [comment]
@candrews #11909 should fix it.

**Comment by Montellese** [comment]
@candrews: as @Rechi mentioned it's already in the works.

Concerning distributions upgrading rapidjson to 1.1.0 I'm not so worried since Ubuntu Zesty is still at 0.12 which is from 2014 and not even a real release ;-) We already have to provide our own PPA build.

**Pull request merged**